xdem/coreg.py

```diff
@@ -1442,3 +1482,457 @@ def _to_matrix_func(self) -> np.ndarray:
+        ])
+        edges_dest = self.apply_pts(edges_source)
+        edges = np.dstack((edges_source, edges_dest))
+
+        all_points = np.append(points, edges, axis=0)
+
+        warped_dem = warp_dem(
+            dem=dem,
+            transform=transform,
+            source_coords=all_points[:, :, 0],
+            destination_coords=all_points[:, :, 1],
+            resampling="linear"
+        )
+
+        return warped_dem
+
+
+
+    def _apply_pts_func(self, coords: np.ndarray) -> np.ndarray:
+        """Apply the scaling model to a set of points."""
+        points = self.to_points()
+
+        new_coords = coords.copy()
+
+        for dim in range(0, 3):
+            with warnings.catch_warnings():
+                # ZeroDivisionErrors may happen when the transformation is empty (which is fine)
+                warnings.filterwarnings("ignore", message="ZeroDivisionError")
+                model = scipy.interpolate.Rbf(
+                    points[:, 0, 0],
+                    points[:, 1, 0],
+                    points[:, dim, 1] - points[:, dim, 0],
+                    function="linear",
+                )
+
+            new_coords[:, dim] += model(coords[:, 0], coords[:, 1])
+
+        return new_coords
+
+
+def warp_dem(
+        dem: np.ndarray,
+        transform: rio.transform.Affine,
+        source_coords: np.ndarray,
+        destination_coords: np.ndarray,
+        resampling: str = "cubic",
+        trim_border: bool = True
+    ) -> np.ndarray:
+    """
+    Warp a DEM using a set of source-destination 2D or 3D coordinates.
+
+    :param dem: The DEM to warp. Allowed shapes are (1, row, col) or (row, col)
+    :param transform: The Affine transform of the DEM.
+    :param source_coords: The source 2D or 3D points. must be X/Y/(Z) coords of shape (N, 2) or (N, 3).
+    :param destination_coords: The destination 2D or 3D points. Must have the exact same shape as 'source_coords'
+    :param resampling: The resampling order to use. Choices: ['nearest', 'linear', 'cubic'].
+    :param trim_border: Remove values outside of the interpolation regime (True) or leave them unmodified (False).
+
+    :raises ValueError: If the inputs are poorly formatted.
+    :raises AssertionError: For unexpected outputs.
+
+    :examples:
+        >>> dem = np.ones((5, 5), dtype=float)
+        >>> dem
+        array([[1., 1., 1., 1., 1.],
+               [1., 1., 1., 1., 1.],
+               [1., 1., 1., 1., 1.],
+               [1., 1., 1., 1., 1.],
+               [1., 1., 1., 1., 1.]])
+        >>> transform = rio.transform.from_origin(0, 5, 1, 1)
+        >>> source_coords = np.array([
+        ...     [0, 0, 0],
+        ...     [5, 0, 0],
+        ...     [5, 5, 0],
+        ...     [0, 5, 0]
+        ... ]).astype("float32")
+        >>> destination_coords = source_coords.copy()
+        >>> destination_coords[3, 2] = 4
+        >>> warped_dem = warp_dem(
```

Comment:
I do not really understand this example...

Reply:
Removed it in ffbf761 as exemplifying it intuitively on that scale doesn't really make sense.